### PR TITLE
Spark-3.5: make `where` sql case sensitive setting alterable in rewrite data files procedure

### PR DESCRIPTION
**Pull request by ludlows**
this pr aims to make the rewriteDataFile action is aware of the user settings about sql case sensitivity in  the  `where` statement. 
 the implementation is simple.
we first obtain the case sensitive setting and save it as a variable in the constructor of rewriteDataFileAction.
then, we pass the variable to the tableScan .

related issue:  https://github.com/apache/iceberg/issues/11438

### CODE REVIEW

**Review by szehon-ho — COMMENTED** [review]
Yes looks reasonable to me as well, agree with @singhpk234 about UT

**Review by anuragmantri — APPROVED** [review]

**Comment by ludlows** [comment]
Hi @szehon-ho ,
please review the test cases, should you have time.
one possible problem is the type of exception is `IllegalArgumentException` here instead of the `ValidationException` mentioned in the issue.

**Comment by ludlows** [comment]
I think I typed the wrong version of iceberg in the issue https://github.com/apache/iceberg/issues/11438

File: spark/v3.5/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestRewriteDataFilesProcedure.java [hunk 1]
```diff
@@ -69,6 +69,60 @@ public void removeTable() {
...
+  @TestTemplate
+  public void testFailsByCaseSensitiveWhereSql() {
+    createTable();
+    insertData(10);
+    sql("set spark.sql.caseSensitive=true");
+    assertThatThrownBy(
+            () ->
+                sql(
+                    "CALL %s.system.rewrite_data_files(table=>'%s', where=>'C1 > 0')",
+                    catalogName, tableIdent))
```
Comment (huaxingao) [inline spark/v3.5/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestRewriteDataFilesProcedure.java:81]: When `spark.sql.caseSensitive` is set to true, this will fail at the Spark analyzer regardless of the changes in this PR.

File: spark/v3.5/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestRewriteDataFilesProcedure.java [hunk 2]
```diff
@@ -69,6 +69,60 @@ public void removeTable() {
...
+
+  @TestTemplate
+  public void testSucceedByCaseInsensitiveWhereSql() {
+    createTable();
+    sql("set spark.sql.caseSensitive=false");
+    assertEquals(
+        "Should have done nothing but passed the schema validation, since no files are present",
+        ImmutableList.of(row(0, 0, 0L, 0)),
+        sql(
+            "CALL %s.system.rewrite_data_files(table=>'%s', where=>'C1 > 0')",
```
Comment (huaxingao) [inline spark/v3.5/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestRewriteDataFilesProcedure.java:94]: With `spark.sql.caseSensitive` set to false, the schema validation will always pass, with or without the changes in this PR.

I think we want to add tests that will fail without the fix and pass with the fix.
We can add a test with `insertData` and `spark.sql.caseSensitive=false`. I think that test will fail without the fix and pass with the fix.

**Comment by huaxingao** [comment]
@ludlows 
Thanks for the PR. I have a couple of questions: 
1. It seems to me that the tests don't really test the changes in this PR; they would pass even without the fix. I think we should add some tests that would fail without the fix but can pass with it. 
3. Do we want to make the Spark SQL configuration `spark.sql.caseSensitive` apply to Iceberg stored procedure parameters? If so, we probably should apply `spark.sql.caseSensitive` to all Iceberg stored procedure parameters. Are there other Iceberg stored procedure parameters that should also honor `spark.sql.caseSensitive`?

**Comment by ludlows** [comment]
1. It seems to me that the tests don't really test the changes in this PR; they would pass even without the fix. I think we should add some tests that would fail without the fix but can pass with it.
2. Do we want to make the Spark SQL configuration spark.sql.caseSensitive apply to Iceberg stored procedure parameters? If so, we probably should apply spark.sql.caseSensitive to all Iceberg stored procedure parameters. Are there other Iceberg stored procedure parameters that should also honor spark.sql.caseSensitive?
----------------------------------
hi  @huaxingao , 
thanks for the questions above.
about the 2nd one, as a data engineer, our users were asking me : why not all parts in the procedure are case insensitive even I have set spark.sql.caseSensitive to false since the procedure is trigger at the sql level but the parameter of procedure is not affected? 

I think it is reasonable to apply the setting about the sql case sensitivity to all procedures, but we could take this pr as the starting point.

File: spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/RewriteDataFilesSparkAction.java [hunk 3]
```diff
@@ -102,11 +102,13 @@ public class RewriteDataFilesSparkAction
   private boolean useStartingSequenceNumber;
   private RewriteJobOrder rewriteJobOrder;
   private FileRewriter<FileScanTask, DataFile> rewriter = null;
+  private boolean caseSensitive;
 
   RewriteDataFilesSparkAction(SparkSession spark, Table table) {
     super(spark.cloneSession());
     // Disable Adaptive Query Execution as this may change the output partitioning of our write
     spark().conf().set(SQLConf.ADAPTIVE_EXECUTION_ENABLED().key(), false);
+    this.caseSensitive = (boolean) spark().conf().get(SQLConf.CASE_SENSITIVE(), true);
```
Comment (huaxingao) [inline spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/RewriteDataFilesSparkAction.java:111]: Can we use `SparkUtil.caseSensitive(spark)` instead?

**Comment by huaxingao** [comment]
@ludlows Thanks for the quick fix. Can we have a test that fails without the fix but passes with it? It seems that all your current tests pass even without the fix.

**Comment by ludlows** [comment]
@huaxingao  I think the test method `testFilterCaseSensitivityBeforeChange()` (leading to validation exception) has shown the bug exists before the PR.

**Comment by huaxingao** [comment]
@ludlows I think you can simply reproduce the problem by something like 
```
    createTable();
    insertData(10);
    sql("SET %s=false", SQLConf.CASE_SENSITIVE().key());
    sql("CALL %s.system.rewrite_data_files(table=>'%s', where=>'C1 > 0'), catalogName, tableIdent));
```

**Comment by ludlows** [comment]
@huaxingao thanks for the comment. 
but i don't think the problem will be raised since the bug has been fixed by this PR. 
please check the test code belove:
```java
@TestTemplate
  public void testFilterCaseSensitivityAfterChange() {
    createTable();
    insertData(10);
    sql("set spark.sql.caseSensitive=false");
    assertEquals(
        "Should have done nothing but passed the schema validation, since no files are present",
        ImmutableList.of(row(0, 0, 0L, 0)),
        sql(
            "CALL %s.system.rewrite_data_files(table=>'%s', where=>'C1 > 90000000')",
            catalogName, tableIdent));
  }
```
the test case above has passed .

**Comment by huaxingao** [comment]
@ludlows Thanks for the quick reply. I know my example will pass with the PR's fix. However, the problem will arise without the fix. We need a simple test that fails without the fix and passes with it. A straightforward test like my example should suffice, with minimal changes. My goal is to keep the test as simple as possible.

File: spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/RewriteDataFilesSparkAction.java [hunk 4]
```diff
@@ -153,6 +156,11 @@ public RewriteDataFilesSparkAction filter(Expression expression) {
     return this;
   }
 
+  public RewriteDataFilesSparkAction setCaseSensitive(boolean caseSensitiveReset) {
```
Comment (huaxingao) [inline spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/RewriteDataFilesSparkAction.java:159]: This method is not used. Remove this unnecessary change?

**Review by huaxingao — APPROVED** [review]
LGTM

**Review by szehon-ho — APPROVED** [review]

**Comment by szehon-ho** [comment]
Thanks @ludlows , and also @huaxingao, @anuragmantri @singhpk234  for reviews
